feat: new iter package

NathanBaulch opened the pull request:
I finally got around to implementing iter support, as discussed last year in #439. This turned into a pretty big task, with over 160 new helper functions based on existing slice oriented helpers in the `iter` sub-package! :sparkles: 

Implementation notes:
1. Requires Go 1.23 - might need to be part of a v2 release. Alternatively it could be a sub-module.
2. I've followed the `I` suffix convention used in the `mutable` package for helper callbacks that include an index parameter.
3. Variadic parameters have been left as is - converting to sequences made helpers awkward to use.
4. `Difference`, `FilterReject`, `Unzip*` not supported - no reasonable way to fan-out to multiple sequences.
5. `Reverse`, `ReduceRight` not supported - would need to buffer entire sequence.
6. `Sample*` not supported - sequence length is unknown and no way to seek randomly forwards/backwards.
7. `DropByIndex`, `Subset`, `*Nth*`, `Splice` don't support negative indexes.
8. `GroupBy`, `PartitionBy`, `Mode` return slices - they already have to scan the full input sequence.
9. `Chunk` returns a sequence of slices - feels awkward returning a sequence of sequences. Can we reuse the allocated slice?
10. `FindUniques*` is implemented but needs to fully iterate the input sequence before yielding anything - not sure if reasonable.
11. `FindDuplicates*` yields items in order of the first duplicate - this order differs slightly from the slice version. 
12. `Fill` is implemented but the name feels misleading - there's no pre-allocated memory being "filled".
13. `DropByIndex`, `WithoutNth` are identical - the only difference between the slice versions is `DropByIndex` supports negative indexes.
14. `DropRight` allocates an internal buffer slice of size n.
15. `DropRightWhile` allocates an internal buffer slice that grows to the longest sub-sequence of items for which the predicate returns true.
16. Map helpers use `iter.Seq2` and do not use `lo.Entry` - more idiomatic.
17. `Entries`, `ToPairs`, `FromEntries`, `FromPairs` accept multiple variadic inputs similar to `Keys` and `Values` - otherwise these helpers wouldn't offer any benefit over `maps.All` and `maps.Collect`.
18. Users will need to be careful about passing infinite sequences to helpers like `lo.Sum` - will block forever.
19. Passing a nil sequence panics - this is consistent with `slices.Collect`.
20. I've used signed `int` or `constraints.Integer` everywhere even if negative arguments are not supported.

samber commented:
Hey @NathanBaulch 

A big thank you for this work. 👏

---

1. Instead of supporting only go >= 1.23, what about a directive similar to the [internal/rand/ package](https://github.com/samber/lo/tree/master/internal/rand) ? I maintain this library as backward compatible as possible, for broad usage and LTS.
3. I agree.
5. + 6. Since you added `GroupBy`+`PartitionBy`+`Mode`, i don't feel uncomfortable adding such helpers, especially if this is properly documented.
11. It does not feel safe to me to reuse the slice. The spirit of this library is to use easy-to-use helpers. If you need to optimize something: build it yourself. This is the Pareto Principle.
18. Good. I have never been satisfied with the `lo.Entry` type. I should have used a `Tuple2`.
22. Good. This is consistent with the rest of the library.

---

I wonder if it could be useful to duplicate some helpers to receive/return `iter.Seq2` iterators. Chaining multiple `lo` helpers without having the `iter.Seq2` variant seems complicated. It can be done in a different PR, but i'm asking this now to prevent any breaking change in method naming.

If we do such a variant, we have to think about the meaning of the 

Anyway, 2 more helpers would be nice: `Seq1ToSeq2` and `Seq2ToSeq1`, with `K` being a forever-increasing `int` in `Seq1ToSeq2`. WDYT?

Other proposal: add a `Void` or `Flush` method for consuming the iterator.

---

I just figured out that `iter` might not be the right name for this package, as suggested in #439. In fact, you cannot import twice a package called "iter" in a go file (`iter` std package + `github.com/samber/lo/iter`).

May I suggest "it"? Any idea?

---

No need to create examples on https://play.golang.org/. First, because we need to merge+release this contribution before, but also because I made an MCP server for creating all of them in 1 prompt.

I think it's time for me to create a dedicated website for documentation. Mixing basic helpers with mutable, parallel, and iterable helpers will make this README.md unreadable.

---

FYI, I will be releasing a reactive programming lib very soon (next week?), with an API similar to rxjs.

The API will be very different and stream-oriented (eg: `Sum` won't return a single value at the end, but as many values sent to it).

NathanBaulch commented on it/channel.go:
```diff
@@ -0,0 +1,29 @@
+import "iter"
+
+// ToChannel returns a read-only channel of collection elements.
+func ToChannel[T any](bufferSize int, collection iter.Seq[T]) <-chan T {
```
What would this return? `<-chan lo.Entry[K, V]` or `<-chan lo.Tuple2[K, V]` maybe?

NathanBaulch commented on iter/type_manipulation.go:
```diff
@@ -0,0 +1,68 @@
+	return func(yield func(T) bool) {
+		for item := range collection {
+			if t, ok := item.(T); !ok {
+				panic("iter.FromAny: type conversion failed")
```
I was originally thinking of aliasing the two stdlib iter types so users only need to import our package, however I don't believe generic type aliases are supported in Go 1.23.
```golang
package iter
type Seq[V any] = iter.Seq[V]
type Seq2[K, V any] = iter.Seq2[K, V]
```

NathanBaulch commented on it/seq.go:
```diff
@@ -0,0 +1,813 @@
+}
+
+// CountValuesBy counts the number of each element returned from mapper function.
+// Is equivalent to chaining lo.Map and lo.CountValues.
```
Fixed.

NathanBaulch commented on it/map.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+// FilterValues transforms a map into a sequence based on predicate returns true for specific elements.
+// It is a mix of lo.Filter() and lo.Values().
```
Fixed.

NathanBaulch commented on it/seq.go:
```diff
@@ -0,0 +1,813 @@
+}
+
+// Keyify returns a map with each unique element of the sequence as a key.
+func Keyify[T comparable](collection iter.Seq[T]) map[T]struct{} {
```
Regarding `Invert`, done!

NathanBaulch commented:
> Since you added `GroupBy`+`PartitionBy`+`Mode`, i don't feel uncomfortable adding such helpers, especially if this is properly documented.

Just to clarify, you're saying I **should** implement `Reverse`, `ReduceRight`, `Sample*`, `Shuffle` (with a disclaimers in the docs) even though they need to scan the full input sequence?

> If we do such a variant, we have to think about the meaning of the

Heh, where was this sentence going?

> 2 more helpers would be nice: `Seq1ToSeq2` and `Seq2ToSeq1`, with `K` being a forever-increasing `int` in `Seq1ToSeq2`.

Would we need two versions of `Seq2ToSeq1`, one for keys and one for values?

> Other proposal: add a `Void` or `Flush` method for consuming the iterator.

I would propose `Drain`.

samber commented:
> Just to clarify, you're saying I should implement Reverse, ReduceRight, Sample*, Shuffle (with a disclaimers in the docs) even though they need to scan the full input sequence?

Yes. Some helpers you built scan the full sequence as well. I would consider that developers understand that reversing a lazily produced slice is not possible without reading it 100%.

> Heh, where was this sentence going?

I don't remember 😅

> Would we need two versions of Seq2ToSeq1, one for keys and one for values?
 
Oh yes, definitely.

> I would propose Drain.

Perfect!

samber commented:
We have too many flaky tests in this project... 🤬

<img width="136" height="289" alt="image" src="https://github.com/user-attachments/assets/e436a1de-5c20-42cd-a015-c794c7bc09ad" />

samber commented:
What do you think about renaming:
- DropRight by DropLast
- ReduceRight by ReduceLast

?

"Left" and "right" are applicable to slices, but not iterators.

In the reactive programming vocabulary, we don't use left/right.

NathanBaulch commented:
Pretty sure I've addressed all outstanding feedback - time for an update!

1. Added `Seq` to the names of helpers corresponding to existing helpers with `Slice` in their name.
2. Replaced `Right` with `Last` and `Left` with `First`.
3. Implemented new helpers `Drain`, `Seq2Channel`, `SeqToSeq2`, `Seq2KeyToSeq`, `Seq2ValueToSeq`.
4. Ported helpers that iterate the full sequence `Reverse`, `ReduceRight`, `Sample`, `Shuffle`.
5. Reviewed all helpers and added documentation if they fully iterate before returning or they allocate internal slices/maps.
6. Ported recent additions `HasPrefix`, `HasSuffix`, `Cut*`, `Trim*`.
   `Cut` isn't supported since it would need to return multiple sequences, similar to `Difference`.
7. Implemented new helper `Length` which corresponds to `len(myslice)`. Can rename to `Len` if preferred.
8. `Intersect` is now variadic, taking any number of input sequences. Perhaps the slice version can also in v2.
9. Renamed `IsSortedByKey` to `IsSortedBy` to be more consistent with other `*By` helpers. Another idea for v2.
10. Renamed all map/filter callback parameters to transform/predicate respectively. "iteratee" is a pretty obscure word and hasn't been used consistently. For example, `lo.ForEachWhile` and `lo.RepeatBy` callbacks are named incorrectly.
11. Significantly reduced duplicate code by using helpers inside helpers. Iterators are well suited to this since they don't allocate return slices.
    For example, `Subset` calls `Slice`, `Replace`, `ToAnySeq` call `Map`, `Compact` calls `Filter`, etc.
12. Lots of performance tweaks, especially around reducing allocations and multiple iteration.
13. `DropByIndex` and `WithoutNth` are identical, `Contains` is the same as `Some` if only passed a single item.

samber commented on .github/workflows/lint.yml:
```diff
@@ -17,6 +17,7 @@ jobs:
       - name: golangci-lint
         uses: golangci/golangci-lint-action@v8
         with:
+          version: 'v2.4'
```
```suggestion
```

mabrarov commented on it/seq.go:
```diff
@@ -0,0 +1,849 @@
+// Associate returns a map containing key-value pairs provided by transform function applied to elements of the given sequence.
+// If any of two pairs have the same key the last one gets added to the map.
+// The order of keys in returned map is not specified and is not guaranteed to be the same from the original sequence.
+func Associate[T any, K comparable, V any](collection iter.Seq[T], transform func(item T) (K, V)) map[K]V {
```
Considering https://github.com/samber/lo/pull/697, don't we need to add into this file:

1. AssociateI
2. SeqToMapI
3. FilterSeqToMap

?

mabrarov replied:
If it will be found reasonable, then here is PR: https://github.com/samber/lo/pull/698

mabrarov commented on it/seq.go:
```diff
@@ -0,0 +1,904 @@
+// If any of two pairs have the same key the last one gets added to the map.
+// The order of keys in returned map is not specified and is not guaranteed to be the same from the original sequence.
+// Will iterate through the entire sequence.
+func Associate[T any, K comparable, V any](collection iter.Seq[T], transform func(item T) (K, V)) map[K]V {
```
If it will be found reasonable, then here is PR: https://github.com/samber/lo/pull/698

samber commented:
Documentation is available here: https://lo.samber.dev/docs/iter/map

The runnable examples will be published on Go playground tomorrow or on Wednesday.
